src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1265,6 +1280,11 @@ void MergeTreeData::dropAllData()
     LOG_TRACE(log, "dropAllData: removing data from memory.");
 
     DataPartsVector all_parts(data_parts_by_info.begin(), data_parts_by_info.end());
+    DataPartsVector committed_parts = getDataPartsVector({DataPartState::Committed});
```

Comment:
Doesn't it involve recurdive mutex locking (see a few lines above)?

Reply:
> Doesn't it involve recursive mutex locking (see a few lines above)?

That did cause a deadlock.  If I use an Ordinary database, the deadlock will definitely occur when dropping table. 

@alexey-milovidov  I have fixed it. PTAL.